### PR TITLE
Cherry pick fact & settings performance improvements to 3.0

### PR DESCRIPTION
**Pull request by ekohl**
This takes [#33639](https://projects.theforeman.org/issues/33639) and [#33640](https://projects.theforeman.org/issues/33640) (with fixes to resolve conflicts) and applies them to 3.0-stable. I've submitted this PR so I can easily patch my own instance to see the differences.

### CODE REVIEW

**Comment by tbrisker** [comment]
@ekohl looks like the test failure might be related

**Comment by ekohl** [comment]
Oh yes, bad cherry pick conflict resolution on my side.

**Comment by tbrisker** [comment]
@ekohl looks like it is still failing 🤔

**Comment by ezr-ondrej** [comment]
> @ekohl looks like it is still failing

It's just because the settings don't see the new shiny category descendant added on the run and do no recache the categories IMHO, so that's not something we should be worried about (we are not adding new setting categories during runtime)

File: test/integration/setting_js_test.rb [hunk 1]
```diff
@@ -23,6 +24,8 @@ def self.humanized_category
       description: 'Pretty setting',
       full_name: 'Pretty setting')
 
+    Foreman.settings.load
```
Comment (jturel) [inline test/integration/setting_js_test.rb]: This change may be unnecessary. Develop had `Foreman.settings.load_definitions` but this branch did not have that call

**Review by tbrisker — APPROVED** [review]
Thanks @ekohl !

**Comment by ekohl** [comment]
In https://github.com/theforeman/foreman/pull/8833 I picked the fact parser improvements to 2.5, but the settings commit had some non-trivial merge conflicts. I'd appreciate it if someone could take a look at that. As can be seen in https://community.theforeman.org/t/foreman-2-5-3-performance-issues-memory-loss-timeouts/25648, it's still a serious issue.

**Comment by jturel** [comment]
I'll get a PR open for a 2.5 pick of that one

**Comment by jturel** [comment]
https://github.com/theforeman/foreman/pull/8834

**Pull request merged**